www/_blog/2021-07-27-storage-beta.mdx

```diff
@@ -0,0 +1,167 @@
+
+![Supabase JS bundle size](/new/images/blog/launch-week-sql-day-2/supabase-js-size.png)
+
+We also considered shipping separate bundles for Node.js and the browser. This would have avoided polyfills for APIs which are already supported in each browser. However, this comes with an increase in complexity to our build process, and users would need to choose which flavor of the library to use in different environments like React Native, Deno, etc. We want to make our libraries universal.
+
+Instead, we modified the storage server to accept Binary data directly. With this change, you can directly upload binary Data and even NodeJS streams to the storage API.
+
+![NodeJS Stream upload](/new/images/blog/launch-week-sql-day-2/stream-upload.png)
+
+With this change in place, the client library supports uploading `ArrayBuffer`, `ArrayBufferView`, `Blob`, `Buffer`, `File`, `FormData`, `NodeJS.ReadableStream`, `ReadableStream<Uint8Array>` - without any change to the bundle size! This should enable you to use storage from an even wider range of environments.
+
+## Context menus
+
+The storage explorer in our Dashboard feels like using a file explorer in your operating system. There were a lot of challenges to make it feel intuitive and robust, since S3 is a not a traditional filesystem. For example, S3 doesn't support a move operation natively - you need to copy the object to the new location and delete the original object. S3 also doesn't have a concept of directory, which means there is no API call to rename a directory. But to make the storage explorer feel like a native file explorer, we implemented these features.
+
+![Context menu 1](/new/images/blog/launch-week-sql-day-2/context-menu-1.png)
+
+![Context menu 2](/new/images/blog/launch-week-sql-day-2/context-menu-2.png)
+
+### The case of the disappearing folders
+
+As we already mentioned, S3 doesn't really have a concept of directories. So if a user uploads an object with path `dir/subdir/avatar.png`, the dir and subdir are automatically "created" and there is no explicit API call to create them.
+
+This leads to an issue where if a user creates an empty directory and refreshes the page, the directory disappears since it is not persisted to S3. To simulate a normal filesystem (where empty directories are not garbage collected), we upload an empty file when the user creates a directory on the dashboard. The dashboard hides this placeholder file and the placeholder is deleted when the user uploads an object in the directory - voila, no more disappearing directores!
+
+## Performance improvements for large buckets
+
+We significantly improved rendering for a large number of items in the Dashboard. We use a virtualization library to render the objects in view and actively discard DOM elements which are not currently rendered from memory. This makes scrolling through thousands of items buttery smooth.
+
+Here is a demo testing rendering the heaviest object known to mankind - `node_modules`!
```

Comment:
Strange that this is appearing like this

<img width="596" alt="image" src="https://user-images.githubusercontent.com/10214025/127175116-75a23770-acf0-4856-982a-6f9e28718f3d.png">